File: runme/mcgen/local/datan/solsrc/s3min.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                         double P_T *r, integer const P_T *n);
extern double CBFUNC minblp(double const P_T *x, integer const P_T *n);
extern double CBFUNC minbsq(double const P_T *x, integer const P_T *n);

/* Common Block Declarations */
/* COMMON /COM/ is needed for Microsoft FORTRAN compiler */
struct {
    double x[2], random[1000], xpl[1000], ypl[1000];
    integer list[2];
    double cx[4] /* was [2][2] */, dxplus[2], dxmins[2],
           scrat[6] /* was [2][3] */, scrat1[4] /* was [2][2] */;
} com;

struct cminbh {
    double t[100], hist[100], deltat;
    integer nt, nevent;
} LIBDATA cminbh;

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__3 = 3, c__4 = 4, c__74 = 74,
               c__1000 = 1000;
static double c_dm10p25 = -10.25, c_dp2 = .2, c_dp9 = .9, c_dp15 = .15,
              c_dp85 = .85, c_dmp414 = -.414, c_d0 = 0., c_dp5 = .5, c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, ich, ipl, ltx, lty, nws, iswit, nseed1, nseed2, lcapt,
                   nstep, nx, ny;
    static double dpl, dx1, dx2, epsiln, fact, fcont, fmin, gamma, rh, dx, dy,
                  xmin, xmax, ymin, ymax, a;
    static char capt[75], tx[75], ty[75];
    static logical ok;
    /* System generated locals */
    integer i__1;
    double d__1, d__2;
/* identify program to user */
    printf("%s\n", "Program S3MIN performs fit of Breit-Wigner function ");
    printf("%s\n\n", "to histogram by minimization.");
/* ask for input */
    printf("%s\n", "Enter number n of events (3 <= n <= 1000):");
    printf("%s", "> ");
    scanf("%li", &cminbh.nevent);
    printf("%s\n", "Enter number (>= 3, <= 100) of histogram bins:");
    printf("%s", "> ");
    scanf("%li", &cminbh.nt);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - likelihood minimization with Poisson stat.");
    printf("%s\n", "2 - minimization of sum of squares");
    printf("%s", "> ");
    scanf("%li", &ich);
/* draw sample */
    nseed1 = 87654;
    nseed2 = 98765;
    rne2in_(&nseed1, &nseed2);
    a = 0.;
    gamma = 1.;
    rnbw(&a, &gamma, com.random, &cminbh.nevent);
/* prepare and fill histogram */
    cminbh.deltat = 20.5 / (double) cminbh.nt;
    i__1 = cminbh.nt;
    for (i = 1; i <= i__1; ++i) {
        cminbh.t[i - 1] = (i - .5) * cminbh.deltat - 10.25;
    }
    smhsin_(cminbh.hist, &c_dm10p25, &cminbh.deltat, &cminbh.nt);
    i__1 = cminbh.nevent;
    for (i = 1; i <= i__1; ++i) {
        smhsfl_(cminbh.hist, &c_dm10p25, &cminbh.deltat, &cminbh.nt,
               &com.random[i - 1], &c_d1);
    }
/* set first approximation */
    com.x[0] = 1.;
    com.x[1] = 2.;
/* minimize with MINSIM */
    com.list[0] = 1;
    com.list[1] = 1;
    nstep = 0;
    epsiln = 0.;
    printf("%s\n", "minimization with MINSIM");
    printf("%s%4li%s%4li%s%4li%s%2li%2li\n", "N = ", cminbh.nt,
           ", NR = ", c__2, ", NRED = ", c__2, ", LIST = ",
           com.list[0], com.list[1]);
    printf("%s%10.5lf%10.5lf\n", "first approx.: X = ", com.x[0], com.x[1]);
    if (ich == 1) {
        minsim_(com.x, &c__2, &c__2, com.list, minblp, &fmin, &epsiln,
               &nstep, com.scrat);
    } else {
        minsim_(com.x, &c__2, &c__2, com.list, minbsq, &fmin, &epsiln,
               &nstep, com.scrat);
    }
    if (nstep < 0) {
        printf("%s\n", "minimization procedure failed");
        exit(0);
    }
    printf("%s%8.2lf%s%6li\n%s%10.5lf%10.5lf\n",
           "result of minimization: FMIN = ", fmin, ", NSTEP =", nstep,
           "X =", com.x[0], com.x[1]);
/* determine covariance matrix */
    fact = 1.;
    if (ich == 1) {
        mincov_(com.x, &c__2, &c__2, com.list, minblp, &fact,
               com.scrat, com.scrat1, com.cx, &ok);
    } else {
        mincov_(com.x, &c__2, &c__2, com.list, minbsq, &fact,
               com.scrat, com.scrat1, com.cx, &ok);
    }
    if (! ok) {
        printf("%s\n", "determination of covariance matrix fails");
        exit(0);
    }
    printf("%s\n", "covariance matrix CX = ");
    mtxwrt_(com.cx, &c__2, &c__2);
    printf("\n");
/* determine asymmetric errors */
    nstep = 0;
    fcont = fmin + .5;
    if (ich == 1) {
        minasy_(minblp, &c__2, &c__2, com.list, com.x, com.cx,
               &fcont, com.dxplus, com.dxmins, com.scrat, &nstep);
    } else {
        minasy_(minbsq, &c__2, &c__2, com.list, com.x, com.cx,
               &fcont, com.dxplus, com.dxmins, com.scrat, &nstep);
    }
    printf("%s\n", "asymmetric errors:");
    printf("%s%10.5lf%10.5lf\n", "DXPLUS = ", com.dxplus[0], com.dxplus[1]);
    printf("%s%10.5lf%10.5lf\n", "DXMINS = ", com.dxmins[0], com.dxmins[1]);
    printf("\n");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
/* plot */
/* prepare caption */
    sprintf(capt, "N = %4li, x_1# = %6.3lf, x_2# = %6.3lf"
            ", &D@x_1# = %6.3lf, &D@x_2# = %6.3lf", cminbh.nevent,
            com.x[0], com.x[1], sqrt(com.cx[0]), sqrt(com.cx[3]));
    lcapt = strlen(capt);
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    strncpy(tx, "y", 1);
    ltx = 1;
    strncpy(ty, "N(y)", 4);
    lty = 4;
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - plot of histogram and fitted function");
    printf("%s\n", "2 - plot of parameters and confidence region");
    printf("%s", "> ");
    scanf("%li", &iswit);
    (void)getchar();
    if (iswit == 1) {
/* draw Gaussian corresponding to solution */
        fact = (double) cminbh.nevent * cminbh.deltat * .6366198;
        dpl = .020520520520520519;
        for (ipl = 1; ipl <= 1000; ++ipl) {
            com.xpl[ipl - 1] = (double) (ipl - 1) * dpl - 10.25;
/* Computing 2nd power */
            d__1 = com.xpl[ipl - 1] - com.x[0];
/* Computing 2nd power */
            d__2 = com.x[1];
            com.ypl[ipl - 1] = fact * abs(com.x[1]) / (d__1 * d__1 * 4. +
                    d__2 * d__2);
        }
        grhscv_(com.xpl, com.ypl, &c__1000, cminbh.hist, &c_dm10p25,
               &cminbh.deltat, &cminbh.nt, tx, &ltx, ty, &lty, capt,
               &lcapt, &nws);
    } else {
/* plot confidence region */
        gropen_();
        gropws_(&nws);
        xmin = com.x[0] - sqrt(com.cx[0]) * 2.;
        xmax = com.x[0] * 2. - xmin;
        ymin = com.x[1] - sqrt(com.cx[3]) * 2.;
        ymax = com.x[1] * 2. - ymin;
        grwncc_(&xmin, &xmax, &ymin, &ymax);
        grvwwc_(&c_dp2, &c_dp9, &c_dp15, &c_dp85);
        grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
        grfram_();
        grboun_();
        grtxtc_(&c_d1, capt, &c__74);
        strncpy(tx, "x_1", 3);
        grsclx_(tx, &c__3);
        strncpy(ty, "x_2", 3);
        ty[3] = ' ';
        grscly_(ty, &c__3);
/* draw solution with symmetric errors and covariance ellipse */
        grstcl_(&c__2);
        dx1 = sqrt(com.cx[0]);
        dx2 = sqrt(com.cx[3]);
        rh = com.cx[2] / (dx1 * dx2);
        if (abs(rh) < .001) {
            rh = .001;
        }
        grdatp_(&c__1, &c_dp5, com.x, &com.x[1], &dx1, &dx2, &rh);
/* draw asymmetric errors */
        grstcl_(&c__3);
        for (i = 1; i <= 2; ++i) {
            if (i == 1) {
                com.xpl[0] = com.x[0] - com.dxmins[0];
            } else {
                com.xpl[0] = com.x[0] + com.dxplus[0];
            }
            com.xpl[1] = com.xpl[0];
            com.ypl[0] = ymin;
            com.ypl[1] = ymax;
            grplin_(&c__2, com.xpl, com.ypl);
        }
        for (i = 1; i <= 2; ++i) {
            if (i == 1) {
                com.ypl[0] = com.x[1] - com.dxmins[1];
            } else {
                com.ypl[0] = com.x[1] + com.dxplus[1];
            }
            com.ypl[1] = com.ypl[0];
            com.xpl[0] = xmin;
            com.xpl[1] = xmax;
            grplin_(&c__2, com.xpl, com.ypl);
        }
        nx = 30;
        ny = 30;
        dx = (xmax - xmin) / nx;
        dy = (ymax - ymin) / ny;
        grstcl_(&c__4);
/* draw confidence region */
        if (ich == 1) {
            mincnt_(&c__1, &c__2, &xmin, &ymin, &dx, &dy, &nx, &ny, &fcont,
                    com.x, &c__2, minblp);
        } else {
            mincnt_(&c__1, &c__2, &xmin, &ymin, &dx, &dy, &nx, &ny, &fcont,
                    com.x, &c__2, minbsq);
        }
        grclse_();
    }
    return 0;
} /* main */

/* ------------------------------------------------------------------ */
void LIBFUNC rnbw(double const P_T *a, double const P_T *gamma,
                  double P_T *r, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;

    /* Local variables */
    static integer i;

    /* Parameter adjustments */
    --r;

    /* Function Body */
    rnecuy_(&r[1], n);
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
        r[i] = *a + *gamma * .5 * tan((r[i] - .5) * 3.14159);
    }
    return;
} /* rnbw */

/* ------------------------------------------------------------------ */
double CBFUNC minblp(double const P_T *x, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;
    double ret_val, d__1, d__2;

    /* Local variables */
    static integer i, ni;
    static double alam, alnlam, alnni, fnorm, gi;

    /* Parameter adjustments */
    --x;

    /* Function Body */
    fnorm = (double) cminbh.nevent * cminbh.deltat;
    ret_val = 0.;
    i__1 = cminbh.nt;
    for (i = 1; i <= i__1; ++i) {
/* GI is the value of the probability density of the population at T(I) */
/* (by replacing the RHS of the following statement it can be */
/*  changed from normal to any desired distribution) */
/* Computing 2nd power */
        d__1 = cminbh.t[i - 1] - x[1];
/* Computing 2nd power */
        d__2 = x[2];
        gi = abs(x[2]) * .6366198 / (d__1 * d__1 * 4. + d__2 * d__2);
/* normalize to number of events in sample */
        gi = fnorm * gi;
        alam = gi;
        if (alam > 0.) {
            ni = nint(cminbh.hist[i - 1]);
            alnlam = log(alam);
            d__1 = (double) (ni + 1);
            alnni = glngam_(&d__1);
            ret_val = ret_val + alnni - cminbh.hist[i - 1] * alnlam + alam;
        }
    }
    return ret_val;
} /* minblp */

/* ------------------------------------------------------------------ */
double CBFUNC minbsq(double const P_T *x, integer const P_T *n)
{
    /* System generated locals */
    integer i__1;
    double ret_val, d__1, d__2;

    /* Local variables */
    static integer i;
    static double fnorm, gi;

    /* Parameter adjustments */
    --x;

    /* Function Body */
    fnorm = (double) cminbh.nevent * cminbh.deltat;
    ret_val = 0.;
    i__1 = cminbh.nt;
    for (i = 1; i <= i__1; ++i) {
/* GI is the value of the probability density of the population at T(I) */
/* (by replacing the RHS of the following statement it can be */
/*  changed from normal to any desired distribution) */
/* Computing 2nd power */
        d__1 = cminbh.t[i - 1] - x[1];
/* Computing 2nd power */
        d__2 = x[2];
        gi = abs(x[2]) * .6366198 / (d__1 * d__1 * 4. + d__2 * d__2);
/* normalize to number of events in sample */
        gi = fnorm * gi;
        if (cminbh.hist[i - 1] > 0.) {
/* Computing 2nd power */
            d__1 = cminbh.hist[i - 1] - gi;
            ret_val += d__1 * d__1 / cminbh.hist[i - 1];
        }
    }
    return ret_val;
} /* minbsq */

File: runme/mcgen/local/datan/solsrc/s3min.f
      PROGRAM S3MIN
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      COMMON /COM/ X(2),RANDOM(1000),XPL(1000),YPL(1000),LIST(2),
     *CX(2,2),DXPLUS(2),DXMINS(2),SCRAT(2,3),SCRAT1(2,2)
C COMMON /COM/ is needed for Microsoft FORTRAN compiler
      COMMON /CMINBH/ T(100),HIST(100),DELTAT,NT,NEVENT
      PARAMETER(NR=2,NRED=2,NPL=1000,T0=-10.25D0)
      DOUBLE PRECISION MINBLP,MINBSQ
      CHARACTER*75 CAPT,TX,TY
      LOGICAL OK
      EXTERNAL MINBLP,MINBSQ
C identify program to user
      WRITE(*,*)' program S3MIN performs fit of Breit-Wigner function'
      WRITE(*,*)' to histogram by minimization'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number n of events (3 <= n <= 1000)'
      WRITE(*,*)' >'
      READ(*,*)NEVENT
      WRITE(*,*)' Enter number (>=3, <=100) of histogram bins'
      WRITE(*,*)' >'
      READ(*,*)NT
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - likelihood minimization with Poisson stat.'
      WRITE(*,*)' 2 - minimization of sum of squares'
      WRITE(*,*)' >'
      READ(*,*)ICH
C draw sample
      NSEED1=87654
      NSEED2=98765
      CALL RNE2IN(NSEED1,NSEED2)
      A=0.D0
      GAMMA=1.D0
      CALL RNBW(A,GAMMA,RANDOM,NEVENT)
C prepare and fill histogram
      DELTAT=2.D0*ABS(T0)/DBLE(NT)
      DO 5 I=1,NT
        T(I)=T0+DBLE(I-0.5D0)*DELTAT
5     CONTINUE
      CALL SMHSIN(HIST,T0,DELTAT,NT)
      DO 10 I=1,NEVENT
        CALL SMHSFL(HIST,T0,DELTAT,NT,RANDOM(I),1.D0)
10    CONTINUE
C set first approximation
      X(1)=1.D0
      X(2)=2.D0
C minimize with MINSIM
      LIST(1)=1
      LIST(2)=1
      NSTEP=0
      EPSILN=0.D0
      WRITE(*,*)' minimization with MINSIM'
      WRITE(*,'(3(A,I4),A,2I2)')' N = ',NT,', NR = ',NR,', NRED = ' ,
     +NRED,', LIST = ',LIST
      WRITE(*,'(A,2F10.5)')' first approx.: X = ',X
      IF(ICH.EQ.1) THEN
        CALL MINSIM(X,NR,NRED,LIST,MINBLP,FMIN,EPSILN,NSTEP,SCRAT)
      ELSE
        CALL MINSIM(X,NR,NRED,LIST,MINBSQ,FMIN,EPSILN,NSTEP,SCRAT)
      END IF
      IF(NSTEP.LT.0) THEN
        WRITE(*,*)' minimization procedure failed'
        STOP
      END IF
      WRITE(*,'(A,F8.2,A,I6,/,A,(2F10.5))')
     +' result of minimization: FMIN = ', FMIN,', NSTEP =',NSTEP,
     +' X =',X
C determine covariance matrix
      FACT=DBLE(1.)
      IF(ICH.EQ.1) THEN
        CALL MINCOV(X,NR,NRED,LIST,MINBLP,FACT,SCRAT,SCRAT1,CX,OK)
      ELSE
        CALL MINCOV(X,NR,NRED,LIST,MINBSQ,FACT,SCRAT,SCRAT1,CX,OK)
      END IF
      IF(.NOT.OK) THEN
        WRITE(*,*)' determination of covariance matrix fails'
        STOP
      END IF
      WRITE(*,*)' covariance matrix CX = '
      CALL MTXWRT(CX,NRED,NRED)
      WRITE(*,'(/)')
C determine asymmetric errors
      NSTEP=0
      FCONT=FMIN+0.5D0
      IF(ICH.EQ.1) THEN
        CALL MINASY(MINBLP,NR,NRED,LIST,X,CX,FCONT,DXPLUS,DXMINS,SCRA
     +  T,NSTEP)
      ELSE
        CALL MINASY(MINBSQ,NR,NRED,LIST,X,CX,FCONT,DXPLUS,DXMINS,SCRA
     +  T,NSTEP)
      END IF
      WRITE(*,*)' asymmetric errors:'
      WRITE(*,'(A,2F10.5)')' DXPLUS = ',DXPLUS
      WRITE(*,'(A,2F10.5)')' DXMINS = ',DXMINS
      WRITE(*,'(/)')
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C plot
C prepare caption
      CAPT(1:38)='N = ****, x_1# = ******, x_2# = ******'
      CAPT(39:74)=', &D@x_1# = ******, &D@x_2# = ******'
      WRITE(CAPT(5:8),'(I4)') NEVENT
      WRITE(CAPT(18:23),'(F6.3)') X(1)
      WRITE(CAPT(33:38),'(F6.3)') X(2)
      WRITE(CAPT(51:56),'(F6.3)') SQRT(CX(1,1))
      WRITE(CAPT(69:74),'(F6.3)') SQRT(CX(2,2))
      LCAPT=74
      TX='y'
      LTX=1
      TY='N(y)'
      LTY=4
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - plot of histogram and fitted function'
      WRITE(*,*)' 2 - plot of parameters and confidence region'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
      IF(ISWIT.EQ.1) THEN
C draw Gaussian corresponding to solution
        FACT=DBLE(NEVENT)*DELTAT*.6366198D0
        DPL=2.D0*ABS(T0)/DBLE(NPL-1)
        DO 30 IPL=1,NPL
          XPL(IPL)=T0+DBLE(IPL-1)*DPL
          YPL(IPL)=FACT*ABS(X(2))/(4.D0*(XPL(IPL)-X(1))**2+X(2)**2)
30      CONTINUE
        CALL GRHSCV(XPL,YPL,NPL,HIST,T0,DELTAT,NT,TX,LTX,TY,LTY,CA
     +  PT,LCAPT,NWS)
      ELSE
C plot confidence region
        CALL GROPEN
        CALL GROPWS(NWS)
        XMIN=X(1)-2.D0*SQRT(CX(1,1))
        XMAX=2.D0*X(1)-XMIN
        YMIN=X(2)-2.D0*SQRT(CX(2,2))
        YMAX=2.D0*X(2)-YMIN
        CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
        CALL GRVWWC(0.2D0,0.9D0,0.15D0,0.85D0)
        CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
        CALL GRFRAM
        CALL GRBOUN
        CALL GRTXTC(1.D0,CAPT,74)
        TX='x_1'
        CALL GRSCLX(TX,3)
        TY='x_2'
        CALL GRSCLY(TY,3)
C draw solution with symmetric errors and covariance ellipse
        CALL GRSTCL(2)
        DX1=SQRT(CX(1,1))
        DX2=SQRT(CX(2,2))
        RH=CX(1,2)/(DX1*DX2)
        IF(ABS(RH).LT..001D0) RH=0.001D0
        CALL GRDATP(1,.5D0,X(1),X(2),DX1,DX2,RH)
C draw asymmetric errors
        CALL GRSTCL(3)
        DO 20 I=1,2
          IF(I.EQ.1) THEN
            XPL(1)=X(1)-DXMINS(1)
          ELSE
            XPL(1)=X(1)+DXPLUS(1)
          END IF
          XPL(2)=XPL(1)
          YPL(1)=YMIN
          YPL(2)=YMAX
          CALL GRPLIN(2,XPL,YPL)
20      CONTINUE
        DO 40 I=1,2
          IF(I.EQ.1) THEN
            YPL(1)=X(2)-DXMINS(2)
          ELSE
            YPL(1)=X(2)+DXPLUS(2)
          END IF
          YPL(2)=YPL(1)
          XPL(1)=XMIN
          XPL(2)=XMAX
          CALL GRPLIN(2,XPL,YPL)
40      CONTINUE
        NX=30
        NY=30
        DX=(XMAX-XMIN)/NX
        DY=(YMAX-YMIN)/NY
        CALL GRSTCL(4)
C draw confidence region
        IF(ICH.EQ.1)THEN
          CALL MINCNT(1,2,XMIN,YMIN,DX,DY,NX,NY,FCONT,X,NR,MINBLP)
        ELSE
          CALL MINCNT(1,2,XMIN,YMIN,DX,DY,NX,NY,FCONT,X,NR,MINBSQ)
        END IF
        CALL GRCLSE
      END IF
      END
C ------------------------------------------------------------------
      SUBROUTINE RNBW(A,GAMMA,R,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(PI=3.14159D0,HALF=0.5D0)
      DIMENSION R(N)
      CALL RNECUY(R,N)
      DO 10 I=1,N
        R(I)=A+HALF*GAMMA*TAN(PI*(R(I)-HALF))
10    CONTINUE
      END
C ------------------------------------------------------------------
      DOUBLE PRECISION FUNCTION MINBLP(X,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      COMMON /CMINBH/ T(100),HIST(100),DELTAT,NT,NEVENT
      DIMENSION X(N)
      PARAMETER(ZERO=0.D0,TOP=.6366198D0)
      FNORM=DBLE(NEVENT)*DELTAT
      MINBLP=ZERO
      DO 10 I=1,NT
C GI is the value of the probability density of the population at T(I)
C (by replacing the RHS of the following statement it can be
C  changed from normal to any desired distribution)
        GI=TOP*ABS(X(2))/(4.D0*(T(I)-X(1))**2+X(2)**2)
C normalize to number of events in sample
        GI=FNORM*GI
        ALAM=GI
        IF(ALAM.GT.ZERO) THEN
          NI=NINT(HIST(I))
            ALNLAM=LOG(ALAM)
            ALNNI=GLNGAM(DBLE(NI+1))
          MINBLP=MINBLP+ALNNI-HIST(I)*ALNLAM+ALAM
        END IF
10    CONTINUE
      END
C ------------------------------------------------------------------
      DOUBLE PRECISION FUNCTION MINBSQ(X,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      COMMON /CMINBH/ T(100),HIST(100),DELTAT,NT,NEVENT
      DIMENSION X(N)
      PARAMETER(ZERO=0.D0,TOP=.6366198D0)
      FNORM=DBLE(NEVENT)*DELTAT
      MINBSQ=ZERO
      DO 10 I=1,NT
C GI is the value of the probability density of the population at T(I)
C (by replacing the RHS of the following statement it can be
C  changed from normal to any desired distribution)
        GI=TOP*ABS(X(2))/(4.D0*(T(I)-X(1))**2+X(2)**2)
C normalize to number of events in sample
        GI=FNORM*GI
        IF(HIST(I).GT.ZERO) THEN
          MINBSQ=MINBSQ+(HIST(I)-GI)**2/HIST(I)
        END IF
10    CONTINUE
      END
